### PR TITLE
graphile config doc overhaul

### PR DESCRIPTION
**Pull request by jcgsville**
## Description

Taking a full pass on the Graphile Config docs. Some highlights:

- Changing the 1-liner to "Graphile Config helps Node.js library authors make their libraries configurable
and _extensible_", which I believe better conveys what the library aims to achieve when compared to "provides a standard plugin interface and helpers that can be used across the entire of the Graphile suite"
- Add a lot more detail around middleware & authoring plugins.
- Reorganize docs around needs of different users: library consumers, plugin authors, and library authors (though I haven't changed much of the content for library authors)

## Checklist

- [x] My code matches the project's code style and `yarn lint:fix` passes.

### CODE REVIEW

File: utils/website/graphile-config/preset.md [hunk 1]
```diff
@@ -1,49 +1,166 @@
...
+export default preset;
+```
+
+:::note
+
+Adding the import statement tells TypeScript about the `GraphileConfig` global
+namespace. No code from the `graphile-config` library will be included in the
+output JavaScript for `graphile.config.ts` above. See the TypeScript docs for
+more info about
+[type-only imports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export).
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:52]:
```suggestion
:::note Importing Graphile Config types is recommended

Adding the `import type {} from "graphile-config"` statement tells TypeScript
about the `GraphileConfig` global namespace; this may or may not be necessary
depending on the other imports that you have, but we generally recommend it.

No code from the `graphile-config` library will be included in the output
JavaScript for `graphile.config.ts` above; see the TypeScript docs for more on
[type-only imports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export).
```

File: utils/website/graphile-config/preset.md [hunk 2]
```diff
@@ -1,49 +1,166 @@
...
+
+Then run your command again.
+
+## Preset scopes
+
+Libraries that use Graphile Config define scopes. Scopes are properties in a
+Graphile Config preset within which configuration options can be set. For
+example, [Postgraphile](https://postgraphile.org/) defines `postgraphile`,
+`pgServices` and other scopes. [Graphile Worker](https://worker.graphile.org/)
+defines a `worker` scope.
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:90]:
```suggestion
example, [PostGraphile](https://postgraphile.org/)'s dependencies define scopes
including `pgServices`, `schema`, `grafast`, `grafserv` and more.
[Graphile Worker](https://worker.graphile.org/) defines a `worker` scope.
```

File: utils/website/graphile-config/preset.md [hunk 3]
```diff
@@ -1,49 +1,166 @@
...
+  extends: [WorkerPreset],
+  worker: {
+    connectionString: "postgres:///my_db",
+  },
+};
+
+export default preset;
+```
+
+:::info
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:119]:
```suggestion
:::info Define your own presets to share your preferred defaults
```

File: utils/website/graphile-config/preset.md [hunk 4]
```diff
@@ -1,49 +1,166 @@
...
+party code, take caution to ensure you can trust the code you are importing.
+
+:::
+
+## Preset resolution
+
+Some libraries that use Graphile Config allow consumers to provide more than one
+preset. Presets may also compose (extend) other presets. When a library is
+passed a list of presets, it resolves the preset using the `ResolvePresets`
+algorithm below.
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:132]: I'm removing this; libraries should only ever import a single preset.

```suggestion
Presets may compose (extend) zero or more other presets. When a library is
passed a preset, it resolves the preset using the `ResolvePreset` algorithm
below.
```

File: utils/website/graphile-config/preset.md [hunk 5]
```diff
@@ -1,49 +1,166 @@
...
+
+tl;dr:
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:134]:
```suggestion
TL;DR:
```

File: utils/website/graphile-config/preset.md [hunk 6]
```diff
@@ -1,49 +1,166 @@
...
+
+- All the presets in `extends` are resolved in order (order is important!).
+- The plugins are merged as a set (each plugin will only be included once) and
+  sorted according to `before`/`after`.
+- The options are merged such that options specified last win.
+
+**ResolvePresets(presets):**
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:141]: Note to self: I'm not 100% sure this algorithm is correct now. Also: we should simplify it.

File: utils/website/graphile-config/preset.md [hunk 7]
```diff
@@ -1,49 +1,166 @@
...
 
-MergePreset(basePreset, extendingPreset):
+**MergePreset(basePreset, extendingPreset):**
 
 1. Let {finalPreset} be an empty preset.
 1. Assert: {basePreset} has an empty or non-existent {extends} property.
-1. Let {plugins} be the list of plugins defined in {basePreset} union the list
-   of plugins in {extendingPreset}.
+1. Let {plugins} be the list of plugins defined in {basePreset} appended with
+   the list of plugins in {extendingPreset}.
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:163]:
```suggestion
1. Let {plugins} be the list of plugins defined in {basePreset} appended with
   those in {extendingPreset} that aren't already present in {basePreset}.
```

File: utils/website/graphile-config/preset.md [hunk 8]
```diff
@@ -58,5 +175,40 @@ MergePreset(basePreset, extendingPreset):
       {extendingScope} actually exist.
 1. Return {finalPreset}.
 
-**IMPORTANT**: the `default` name must not be used as a top-level key in a
-preset to enable compatibility with the various ESM emulations.
+:::info
+
+If the following are true:
+
+1. You provide two presets - PresetA and PresetB - to a library.
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:182]:
```suggestion
1. A preset extends two other presets - PresetA and PresetB.
```

File: utils/website/graphile-config/preset.md [hunk 9]
```diff
@@ -58,5 +175,40 @@ MergePreset(basePreset, extendingPreset):
...
+
+:::
+
+## `graphile` CLI
+
+:::tip
+
+The `graphile` CLI is only available to sponsors. See the
+[README](https://github.com/graphile/crystal/blob/main/utils/graphile/README.md)
+for more details.
```
Comment (benjie) [inline utils/website/graphile-config/preset.md:202]:
```suggestion
:::tip Sponsorship has perks: the `graphile` CLI!

Graphile's open source projects only exist thanks to the sponsorship of
individuals and organizations that use them. To help convince your boss to fund
the ongoing development of the OSS that your company relies on, sponsorship
comes with perks. One such perk is the `graphile` CLI which is only licensed for
usage by sponsors. See the
[README](https://github.com/graphile/crystal/blob/main/utils/graphile/README.md)
for more details.
```

File: utils/website/src/components/HomepageFeatures/index.js [hunk 10]
```diff
@@ -10,8 +10,8 @@ const FeatureList = [
     link: "/graphile-config/",
     description: (
       <>
-        Provides a standard plugin interface and helpers which can be used
-        across the entire Graphile suite
+        Helps Node.js library authors make their libraries configurable and
```
Comment (benjie) [inline utils/website/src/components/HomepageFeatures/index.js:13]:
```suggestion
        Helps Node.js module authors make their libraries configurable and
```

**Review by benjie — CHANGES_REQUESTED** [review]
This is great! Thanks so much for your work on this! :raised_hands: 

I've done a lot of editorial, but it's mostly just small edits for for clarity and correctness, or to align with my general writing tone.

I think we should reconsider restoring the middleware page though - plugins aren't just middleware (e.g. Graphile-Build plugins also have inflection, behaviors, gather, schema hooks, etc etc), middleware is just one of the standardized features that plugins offer. In future, I would like to add more dedicated pages for other features (e.g. if we standardize inflection across all our libraries) rather than making the plugin page longer...

Actually what would probably make sense would be to move `plugin.md` to `plugins/index.md` and then create `plugins/middleware.md` as well - group them together under the "plugin" banner, but still a dedicated page.

File: utils/website/graphile-config/plugin.md [hunk 11]
```diff
@@ -1,43 +1,381 @@
...
+```ts
+await middleware.run(
+  "someAction", // The "action name"
+  { someParameter: 42 }, // The `event` object
+  async (event) => {
+    // The "underlying action"
+    // This function is what will be retried if the MySomeActionRetryPlugin is
+    // included in a preset.
+  },
+);
```
[the remark this answers is not in the document]
Comment (jcgsville) [reply utils/website/graphile-config/plugin.md:169]: On first glance, I don't agree with your suggestion here. It suggests that doTheThing() will be retried and is the underlying action, but I'd argue that the anonymous function that starts with `async(event) => {` is the underlying action.

[the remark this answers is not in the document]
Comment (jcgsville) [reply utils/website/graphile-config/plugin.md:169]: What about

```ts
const actionResult = await middleware.run(
  "someAction", // The "action name"
  { someParameter: 42 }, // The `event` object
  // The "underlying action"
  // This function is what will be retried if the MySomeActionRetryPlugin is
  // included in a preset.
  async (event) => {
    // Extract the (possibly modified) values from the event
    const { someParameter } = event;
    return doTheThing(someParameter);
  },
);
```

File: utils/website/graphile-config/plugin.md [hunk 12]
```diff
@@ -1,43 +1,381 @@
...
+
+export const MySendgridPlugin: GraphileConfig.Plugin = {
+  name: "MySendgridPlugin",
+  libraryName: {
+    middleware: {
+      foo(next, event) {
+        new SendgridSdk(
+          // This assumes that the library passes in the resolved preset via
+          // the event object. This will vary depending on the library.
+          event.ctx.resolvedPreset.sendgrid.apiKey,
```
[the remark this answers is not in the document]
Comment (jcgsville) [reply utils/website/graphile-config/plugin.md]: Used your suggestion and updated the comment 🙂

**Comment by jcgsville** [comment]
Thanks for the thoughtful feedback

**Comment by jcgsville** [comment]
This should be ready unless you want to do something about the ResolvePreset pseudocode

https://github.com/graphile/crystal/pull/2282#discussion_r1890307759

**Review by benjie — APPROVED** [review]
Awesome; thanks for your hard work on this! :raised_hands:
